aptos-move/aptos-e2e-comparison-testing/src/execution.rs

```diff
@@ -0,0 +1,358 @@
+            features.features[byte_index] |= bit_mask;
+        }
+    }
+
+    pub async fn execute_txns(&self, begin: Version, limit: u64) -> Result<()> {
+        let aptos_commons_path = self.input_path.join(APTOS_COMMONS);
+        if !check_aptos_packages_availability(aptos_commons_path.clone()) {
+            return Err(anyhow::Error::msg("aptos packages are missing"));
+        }
+
+        let mut compiled_package_cache: HashMap<PackageInfo, HashMap<ModuleId, Vec<u8>>> =
+            HashMap::new();
+        let mut compiled_package_cache_v2: HashMap<PackageInfo, HashMap<ModuleId, Vec<u8>>> =
+            HashMap::new();
+        if self.execution_mode.is_v1() || self.execution_mode.is_compare() {
+            compile_aptos_packages(&aptos_commons_path, &mut compiled_package_cache, false)?;
+        }
+        if self.execution_mode.is_v2() || self.execution_mode.is_compare() {
+            compile_aptos_packages(&aptos_commons_path, &mut compiled_package_cache_v2, true)?;
+        }
+
+        // prepare data
+        let data_manager = DataManager::new(&self.input_path);
+        if !data_manager.check_dir_availability() {
+            return Err(anyhow::Error::msg("data is missing"));
+        }
+        if !IndexReader::check_availability(&self.input_path) {
+            return Err(anyhow::Error::msg("index file is missing"));
+        }
+        let mut index_reader = IndexReader::new(&self.input_path);
+
+        // get the first idx from the version_index file
+        let ver = index_reader.get_next_version_ge(begin);
+        if ver.is_none() {
+            return Err(anyhow::Error::msg(
+                "cannot find a version greater than or equal to the specified begin version",
+            ));
+        }
+        let mut cur_version = ver.unwrap();
+        while cur_version < begin + limit {
+            self.execute_one_txn(
+                cur_version,
+                &data_manager,
+                &mut compiled_package_cache,
+                &mut compiled_package_cache_v2,
+            )?;
+            if let Some(ver) = index_reader.get_next_version() {
+                cur_version = ver;
+            } else {
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    fn compile_code(
+        &self,
+        txn_idx: &TxnIndex,
+        compiled_package_cache: &mut HashMap<PackageInfo, HashMap<ModuleId, Vec<u8>>>,
+        compiled_package_cache_v2: &mut HashMap<PackageInfo, HashMap<ModuleId, Vec<u8>>>,
+    ) -> Result<()> {
+        if !txn_idx.package_info.is_compilable() {
+            return Err(anyhow::Error::msg("not compilable"));
+        }
+        let package_info = txn_idx.package_info.clone();
+        let package_dir = self.input_path.join(format!("{}", package_info));
+        if !package_dir.exists() {
+            return Err(anyhow::Error::msg("source code is not available"));
+        }
+        if (self.execution_mode.is_compare() || self.execution_mode.is_v1())
```

Comment:
nit: we repeated this a few times, how about adding an `execution_mode.needs_v1()`?

Reply:
done